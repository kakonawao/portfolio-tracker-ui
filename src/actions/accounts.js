import fetch from "cross-fetch";

import { getRequestOptions, handleResponse } from "./util";
import { getEndpoint, PATHS } from "../config";


export const REQUEST_ACCOUNTS = "REQUEST_ACCOUNTS";
export const FAIL_ACCOUNTS = "FAIL_ACCOUNTS";
export const RECEIVE_ACCOUNTS = "RECEIVE_ACCOUNTS";

export const requestAccounts = () => {
    return {
        type: REQUEST_ACCOUNTS
    }
};

export const failAccounts = (error) => {
    return {
        type: FAIL_ACCOUNTS,
        error
    }
};

export const receiveAccounts = (data) => {
    return {
        type: RECEIVE_ACCOUNTS,
        data
    }
};

export const fetchAccounts = (session) => {
    return function (dispatch) {
        dispatch(requestAccounts());

        return fetch(getEndpoint(PATHS.ACCOUNTS), getRequestOptions(session))
            .then(response => handleResponse(response))
            .then(data => dispatch(receiveAccounts(data)))
            .catch(error => dispatch(failAccounts(error)));
    }
};
