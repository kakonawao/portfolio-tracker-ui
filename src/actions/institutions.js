import fetch from "cross-fetch";

import {getRequestOptions, handleResponse} from "./util";
import { getEndpoint, PATHS } from "../config";

export const REQUEST_INSTITUTIONS = "REQUEST_INSTITUTIONS";
export const RECEIVE_INSTITUTIONS = "RECEIVE_INSTITUTIONS";
export const FAIL_INSTITUTIONS = "FAIL_INSTITUTIONS";


export const requestInstitutions = () => {
    return {
        type: REQUEST_INSTITUTIONS
    }
};

export const failInstitutions = (error) => {
    return {
        type: FAIL_INSTITUTIONS,
        error
    }
};

export const receiveInstitutions = (data) => {
    return {
        type: RECEIVE_INSTITUTIONS,
        data
    }
};

export const fetchInstitutions = (session) => {
    return function (dispatch) {

        dispatch(requestInstitutions());

        return fetch(getEndpoint(PATHS.INSTITUTIONS), getRequestOptions(session))
            .then(response => handleResponse(response))
            .then(data => dispatch(receiveInstitutions(data)))
            .catch(error => dispatch(failInstitutions(error)));
    }
};
