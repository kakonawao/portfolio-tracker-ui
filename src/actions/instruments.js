import fetch from "cross-fetch";

import {getRequestOptions, handleResponse} from "./util";
import { getEndpoint, PATHS } from "../config";


export const REQUEST_INSTRUMENTS = "REQUEST_INSTRUMENTS";
export const RECEIVE_INSTRUMENTS = "RECEIVE_INSTRUMENTS";
export const FAIL_INSTRUMENTS = "FAIL_INSTRUMENTS";

export const requestInstruments = () => {
    return {
        type: REQUEST_INSTRUMENTS
    }
};

export const failInstruments = (error) => {
    return {
        type: FAIL_INSTRUMENTS,
        error
    }
};

export const receiveInstruments = (data) => {
    return {
        type: RECEIVE_INSTRUMENTS,
        data
    }
};

export const fetchInstruments = (session) => {
    return function (dispatch) {

        dispatch(requestInstruments());

        return fetch(getEndpoint(PATHS.INSTRUMENTS), getRequestOptions(session))
            .then(response => handleResponse(response))
            .then(data => dispatch(receiveInstruments(data)))
            .catch(error => dispatch(failInstruments(error)));
    }
};
