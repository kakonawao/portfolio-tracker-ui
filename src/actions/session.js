import fetch from "cross-fetch";

import { getRequestOptions, handleResponse } from "./util";
import { getEndpoint, PATHS } from "../config";


export const REQUEST_TOKEN = "REQUEST_TOKEN";
export const RECEIVE_TOKEN = "RECEIVE_TOKEN";
export const FAIL_TOKEN = "FAIL_TOKEN";

export const requestToken = () => {
    return {
        type: REQUEST_TOKEN
    };
};


export const receiveToken = (data) => {
    return {
        type: RECEIVE_TOKEN,
        data
    }
};

export const failToken = (error) => {
    return {
        type: FAIL_TOKEN,
        error
    }
};

export const login = (data) => {
    return function (dispatch) {
        dispatch(requestToken());

        const formData = {
            ...data,
            grant_type: "password"
        };
        const body = new URLSearchParams(formData).toString();
        const url = getEndpoint(PATHS.SESSION);
        const opts = getRequestOptions(null, "POST", body, "x-www-form-urlencoded");

        return fetch(url, opts)
            .then(response => handleResponse(response))
            .then(data => dispatch(receiveToken({
                    username: data.username,
                    token: data.access_token,
                    tokenType: data.token_type})))
            .catch(error => dispatch(failToken(error)));
    }
};
