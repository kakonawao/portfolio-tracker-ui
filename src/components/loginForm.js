import React from "react";


class LoginForm extends React.Component {

    render() {
        return (
            <form>
                <div>
                    <label>
                        Username
                        <input type={"text"}
                               name={"username"}
                               ref={this.props.username} />
                    </label>
                    <label>
                        Password
                        <input type={"password"}
                               name={"password"}
                               ref={this.props.password} />
                    </label>
                </div>
                <div>
                    <button type={"submit"}
                            onClick={this.props.login}>
                        Login
                    </button>
                </div>
            </form>
        );
    }
}

export default LoginForm;
