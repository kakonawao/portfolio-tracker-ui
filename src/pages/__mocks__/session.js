export const login = jest.fn();
