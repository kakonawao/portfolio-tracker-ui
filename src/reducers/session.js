import {
    FAIL_TOKEN,
    RECEIVE_TOKEN,
    REQUEST_TOKEN
} from "../actions/session";


const initialState = {
    loading: false,
    error: null,
    authenticated: false,
    username: null,
    token: null,
    tokenType: null
};

export const sessionReducer = (state = initialState, action) => {
    switch(action.type) {
        case REQUEST_TOKEN:
            return {
                ...initialState,
                loading: true
            };
        case FAIL_TOKEN:
            return {
                ...state,
                loading: false,
                error: action.error
            };
        case RECEIVE_TOKEN:
            return {
                ...state,
                ...action.data,
                loading: false,
                authenticated: true
            };
        default:
            return state;
    }
};
