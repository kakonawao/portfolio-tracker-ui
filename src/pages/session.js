import React from "react";
import { connect } from "react-redux";

import LoginForm from "../components/loginForm";
import { login } from "../actions/session";


const mapStateToProps = (state) => {
    return {
        authenticated: state.session.authenticated,
        username: state.session.username
    }
};

export class SessionPage extends React.Component {

    constructor(props) {
        super(props);

        this.inputs = {
            username: React.createRef(),
            password: React.createRef()
        }
    }

    login = (event) => {
        event.preventDefault();
        const loginData = {
            username: this.inputs.username.current.value,
            password: this.inputs.password.current.value
        };
        this.props.dispatch(login(loginData));
    };

    renderUserData() {
        return (
            <div>
                <h2>Session</h2>
                <p>You are logged in as {this.props.username}.</p>
            </div>
        )
    }

    renderLoginForm() {
        return (
            <div>
                <h2>Session</h2>
                <LoginForm login={this.login}
                           username={this.inputs.username}
                           password={this.inputs.password}/>
            </div>
        );
    }

    render() {
        if (this.props.authenticated) {
            return this.renderUserData();
        }

        return this.renderLoginForm();
    };
}

export default connect(mapStateToProps)(SessionPage);
