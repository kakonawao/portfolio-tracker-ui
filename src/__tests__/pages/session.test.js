import React from "react";
import { shallow } from "enzyme";

import { SessionPage } from "../../pages/session";
import { LoginForm } from "../../components/loginForm";
import { login } from "../../actions/session";

jest.mock("../../actions/session");


describe("<SessionPage>", () => {

    it("user can log in when user is not authenticated", () => {
        // Shallow-render
        const props = {
            dispatch: jest.fn(),
            autenticated: false,
            username: null
        };
        const wrapper = shallow(<SessionPage {...props}/>);

        // Validate snapshot
        expect(wrapper).toMatchSnapshot();

        // Simulate login action
        wrapper.instance().inputs.username.current = {value: "peter"};
        wrapper.instance().inputs.password.current = {value: "6r1ff1n"};
        const ev = { preventDefault: jest.fn()};
        wrapper.find("LoginForm").props().login(ev);

        // Validate login action was dispatched correctly
        expect(ev.preventDefault.mock.calls.length).toBe(1);
        expect(login.mock.calls[0][0]).toEqual({username: "peter", password: "6r1ff1n"});
        expect(props.dispatch.mock.calls.length).toBe(1);
    });

    it("renders username when user is authenticated", () => {
        // Shallow-render
        const props = {
            dispatch: jest.fn(),
            authenticated: true,
            username: "potato"
        };
        const wrapper = shallow(<SessionPage {...props}/>);

        // Validate snapshot
        expect(wrapper).toMatchSnapshot();
    });
});
