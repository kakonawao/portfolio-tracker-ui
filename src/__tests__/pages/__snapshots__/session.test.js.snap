// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SessionPage> renders username when user is authenticated 1`] = `
<div>
  <h2>
    Session
  </h2>
  <p>
    You are logged in as 
    potato
    .
  </p>
</div>
`;

exports[`<SessionPage> user can log in when user is not authenticated 1`] = `
<div>
  <h2>
    Session
  </h2>
  <LoginForm
    login={[Function]}
    password={
      Object {
        "current": null,
      }
    }
    username={
      Object {
        "current": null,
      }
    }
  />
</div>
`;
